File: androidApp/src/main/java/io/lb/middleware/android/createroute/presentation/originalroute/OriginalRouteScreen.kt
package io.lb.middleware.android.createroute.presentation.originalroute

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavHostController
import io.lb.middleware.android.core.presentation.Screens
import io.lb.middleware.android.core.presentation.components.DefaultTextButton
import io.lb.middleware.android.core.presentation.components.DefaultTextField
import io.lb.middleware.android.core.presentation.components.GenericTopAppBar
import io.lb.middleware.android.core.presentation.components.MapElement
import io.lb.middleware.android.core.presentation.components.MethodBox
import io.lb.middleware.android.core.presentation.components.TestColumn
import io.lb.middleware.android.core.presentation.showToast
import io.lb.middleware.android.core.util.generateOldBodyFieldsFromJson
import io.lb.middleware.android.createroute.presentation.model.AndroidOldBodyField
import io.lb.middleware.android.createroute.presentation.model.CreateRouteArgs
import io.lb.middleware.common.shared.middleware.request.MiddlewareHttpMethods
import io.lb.middleware.common.state.CommonFlow
import io.lb.middleware.shared.presentation.createroute.originalroute.OriginalRouteEvent
import io.lb.middleware.shared.presentation.createroute.originalroute.OriginalRouteState
import io.lb.middleware.shared.presentation.createroute.originalroute.OriginalRouteViewModel
import kotlinx.coroutines.flow.collectLatest

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun OriginalRouteScreen(
    navController: NavHostController,
    state: OriginalRouteState,
    eventFlow: CommonFlow<OriginalRouteViewModel.UiEvent>,
    onEvent: (OriginalRouteEvent) -> Unit
) {
    val context = LocalContext.current
    val result = remember {
        mutableStateOf("")
    }
    val code = remember {
        mutableIntStateOf(0)
    }
    val originalBaseUrl = remember {
        mutableStateOf("https://")
    }
    val originalPath = remember {
        mutableStateOf("")
    }
    val originalMethodExpanded = remember {
        mutableStateOf(false)
    }
    val originalMethod = remember {
        mutableStateOf(MiddlewareHttpMethods.Get)
    }
    val originalBody = remember {
        mutableStateOf("")
    }

    LaunchedEffect(key1 = Screens.ORIGINAL_ROUTE) {
        eventFlow.collectLatest {
            when (it) {
                is OriginalRouteViewModel.UiEvent.ShowError -> {
                    context.showToast(it.message)
                }
                is OriginalRouteViewModel.UiEvent.ShowOriginalRouteSuccess -> {
                    context.showToast("Called original route successfully!")
                }

                is OriginalRouteViewModel.UiEvent.ShowResult -> {
                    code.intValue = it.code
                    result.value = it.result
                }

                OriginalRouteViewModel.UiEvent.NavigateToNextStep -> {
                    val oldFields = generateOldBodyFieldsFromJson(result.value)
                    val args = CreateRouteArgs(
                        originalResponse = result.value,
                        originalBaseUrl = originalBaseUrl.value,
                        originalPath = originalPath.value,
                        originalMethod = originalMethod.value,
                        originalBody = originalBody.value,
                        originalQueries = state.originalQueries,
                        originalHeaders = state.originalHeaders,
                        oldBodyFields = oldFields.mapValues { old ->
                            AndroidOldBodyField.fromOldBodyField(old.value)
                        }
                    )
                    navController.currentBackStackEntry?.arguments?.putParcelable("CreateRouteArgs", args)
                    navController.navigate(Screens.FILL_ROUTE_FIELDS.name)
                }
            }
        }
    }

    Scaffold(
        modifier = Modifier.fillMaxSize(),
        topBar = {
            GenericTopAppBar(navController, "Step 1: Original Route")
        },
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .padding(padding)
                .fillMaxSize()
                .padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.Top,
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            item {
                DefaultTextButton(
                    modifier = Modifier.fillMaxWidth()
                        .padding(
                            vertical = 8.dp,
                            horizontal = 32.dp
                        ),
                    text = "Move Forward",
                    containerColor = MaterialTheme.colorScheme.surface,
                    contentColor = MaterialTheme.colorScheme.onSurface,
                    enabled = state.isLoading.not(),
                    onClick = {
                        onEvent(
                            OriginalRouteEvent.MoveForward(
                                result = result.value,
                                code = code.intValue,
                                originalBaseUrl = originalBaseUrl.value,
                                originalPath = originalPath.value,
                                originalMethod = originalMethod.value,
                                originalBody = originalBody.value
                            )
                        )
                    }
                )
            }

            item {
                Box(
                    modifier = Modifier.fillMaxWidth()
                ) {
                    MethodBox(
                        modifier = Modifier
                            .fillMaxWidth(0.2f)
                            .clickable {
                                originalMethodExpanded.value = true
                            },
                        method = originalMethod.value,
                        text = "Original Method"
                    )
                    DropdownMenu(
                        expanded = originalMethodExpanded.value,
                        onDismissRequest = { originalMethodExpanded.value = false }
                    ) {
                        MiddlewareHttpMethods.entries.forEach {
                            DropdownMenuItem(
                                text = { MethodBox(it) },
                                onClick = {
                                    originalMethod.value = it
                                    originalMethodExpanded.value = false
                                }
                            )
                        }
                    }
                }
            }

            item {
                DefaultTextField(
                    modifier = Modifier
                        .padding(vertical = 8.dp)
                        .fillMaxWidth(),
                    text = originalBaseUrl.value,
                    isEnabled = state.isLoading.not(),
                    label = "Original Base URL",
                    onValueChange = {
                        originalBaseUrl.value = it
                    }
                )
            }

            item {
                DefaultTextField(
                    modifier = Modifier
                        .padding(vertical = 8.dp)
                        .fillMaxWidth(),
                    text = originalPath.value,
                    isEnabled = state.isLoading.not(),
                    label = "Original Path",
                    onValueChange = {
                        originalPath.value = it
                    }
                )
            }

            item {
                DefaultTextField(
                    modifier = Modifier
                        .padding(vertical = 8.dp)
                        .fillMaxWidth(),
                    text = originalBody.value,
                    isEnabled = state.isLoading.not(),
                    label = "Original Body",
                    isSingleLined = false,
                    onValueChange = {
                        originalBody.value = it
                    }
                )
            }

            item {
                Spacer(modifier = Modifier.padding(6.dp))
            }

            item {
                Text(
                    modifier = Modifier.fillMaxWidth(),
                    text = "Original Headers",
                    fontSize = 16.sp,
                    color = MaterialTheme.colorScheme.primary
                )
            }

            items(state.originalHeaders.keys.toList()) {
                MapElement(
                    key = it,
                    value = state.originalHeaders[it] ?: return@items,
                    isLoading = state.isLoading,
                    isAdded = state.originalHeaders.containsKey(it),
                    onClickAdd = { key, value ->
                        onEvent(OriginalRouteEvent.UpsertOriginalHeader(key, value))
                    },
                    onClickRemove = { key ->
                        onEvent(OriginalRouteEvent.RemoveOriginalHeader(key))
                    },
                )
            }

            item {
                MapElement(
                    isLoading = state.isLoading,
                    isAdded = false,
                    onClickAdd = { key, value ->
                        onEvent(OriginalRouteEvent.UpsertOriginalHeader(key, value))
                    },
                    onClickRemove = { key ->
                        onEvent(OriginalRouteEvent.RemoveOriginalHeader(key))
                    },
                )
            }

            item {
                Spacer(modifier = Modifier.padding(6.dp))
            }

            item {
                Text(
                    modifier = Modifier.fillMaxWidth(),
                    text = "Original Queries",
                    fontSize = 16.sp,
                    color = MaterialTheme.colorScheme.primary
                )
            }

            items(state.originalQueries.keys.toList()) {
                MapElement(
                    key = it,
                    value = state.originalQueries[it] ?: return@items,
                    isLoading = state.isLoading,
                    isAdded = state.originalQueries.containsKey(it),
                    onClickAdd = { key, value ->
                        onEvent(OriginalRouteEvent.UpsertOriginalQuery(key, value))
                    },
                    onClickRemove = { key ->
                        onEvent(OriginalRouteEvent.RemoveOriginalQuery(key))
                    },
                )
            }

            item {
                MapElement(
                    isLoading = state.isLoading,
                    isAdded = false,
                    onClickAdd = { key, value ->
                        onEvent(OriginalRouteEvent.UpsertOriginalQuery(key, value))
                    },
                    onClickRemove = { key ->
                        onEvent(OriginalRouteEvent.RemoveOriginalQuery(key))
                    },
                )
            }

            item {
                TestColumn(
                    isLoading = state.isLoading,
                    result = if (result.value.isNotBlank()) {
                        "Code: ${code.intValue}\nBody:\n${result.value}"
                    } else {
                        ""
                    },
                ) {
                    result.value = ""
                    onEvent(
                        OriginalRouteEvent.TestOriginalRoute(
                            originalBaseUrl = originalBaseUrl.value,
                            originalPath = originalPath.value,
                            originalMethod = originalMethod.value,
                            originalBody = originalBody.value
                        )
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/create_route/original_route/OriginalRouteScreen.swift
//
//  OriginalRouteScreen.swift
//  iosApp
//
//  Created by Leonardo Bai on 06/04/25.
//  Copyright © 2025 orgName. All rights reserved.
//

import SwiftUI
import Shared

struct OriginalRouteScreen: View {
    private var testOriginalRouteUseCase: Middleware_domainTestOriginalRouteUseCase
    @ObservedObject private var viewModel: IOSOriginalRouteViewModel
    
    @State private var result = ""
    @State private var code = 0
    @State private var originalBaseUrl = "https://"
    @State private var originalPath = ""
    @State private var originalMethodExpanded = false
    @State private var originalMethod = Common_sharedMiddlewareHttpMethods.get
    @State private var originalBody = ""
    
    @State private var showErrorAlert = false
    @State private var errorMessage = ""
    @State private var eventSubscription: StateDisposableHandle?
    @Environment(\.navigate) private var navigate
    
    init(
        testOriginalRouteUseCase: Middleware_domainTestOriginalRouteUseCase
    ) {
        self.testOriginalRouteUseCase = testOriginalRouteUseCase
        self.viewModel = IOSOriginalRouteViewModel(
            testOriginalRouteUseCase: testOriginalRouteUseCase
        )
    }
    
    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                methodSelectionSection
                inputFieldsSection
                headersSection
                queriesSection
                testSection
            }
            .padding(.vertical)
        }
        .onAppear { setupScreen() }
        .onDisappear { cleanup() }
        .background(Color.background)
        .toolbar {
            ToolbarItem(placement: .navigationBarTrailing) {
                Button(
                    action: {
                        viewModel.onEvent(event: OriginalRouteEvent.MoveForward(
                            result: result,
                            code: Int32(code),
                            originalBaseUrl: originalBaseUrl,
                            originalPath: originalPath,
                            originalMethod: originalMethod,
                            originalBody: originalBody
                        ))
                    }
                ) {
                    HStack {
                        Text("Next")
                            .foregroundColor(Color.primaryColor)
                        Image(systemName: "chevron.right")
                            .foregroundColor(Color.primaryColor)
                    }
                }
            }
        }
        .alert("Error", isPresented: $showErrorAlert) {
            Button("OK", role: .cancel) { }
        } message: {
            Text(errorMessage)
        }
        .navigationTitle("Step 1/5: Original Route")
    }
    
    // MARK: - Subviews
    
    private var methodSelectionSection: some View {
        HStack {
            MethodBox(
                method: originalMethod,
                text: "Original Method"
            )
            .onTapGesture {
                originalMethodExpanded = true
            }
            .confirmationDialog(
                "Select Method",
                isPresented: $originalMethodExpanded,
                titleVisibility: .visible
            ) {
                ForEach(Common_sharedMiddlewareHttpMethods.entries, id: \.self) { method in
                    Button(method.name) {
                        originalMethod = method
                    }
                }
            }
            Spacer()
        }
        .padding(.horizontal)
    }
    
    private var inputFieldsSection: some View {
        VStack(spacing: 8) {
            DefaultTextField(
                text: $originalBaseUrl,
                label: "Original Base URL",
                isEnabled: !viewModel.state.isLoading
            )
            
            DefaultTextField(
                text: $originalPath,
                label: "Original Path",
                isEnabled: !viewModel.state.isLoading
            )
            
            DefaultTextField(
                text: $originalBody,
                label: "Original Body",
                isEnabled: !viewModel.state.isLoading,
                isSingleLined: false
            )
        }
        .padding(.horizontal)
    }
    
    private var headersSection: some View {
        Group {
            Text("Original Headers")
                .font(.headline)
                .foregroundColor(Color.primaryColor)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal)
            
            // Existing headers
            ForEach(Array(viewModel.state.originalHeaders.keys.sorted()), id: \.self) { key in
                MapElement(
                    key: key,
                    value: viewModel.state.originalHeaders[key] ?? "",
                    isLoading: viewModel.state.isLoading,
                    isAdded: true,
                    onClickAdd: { key, value in
                        viewModel.onEvent(event: OriginalRouteEvent.UpsertOriginalHeader(key: key, value: value))
                    },
                    onClickRemove: { key in
                        viewModel.onEvent(event: OriginalRouteEvent.RemoveOriginalHeader(key: key))
                    }
                )
                .padding(.horizontal)
            }
            
            // New header input
            MapElement(
                isLoading: viewModel.state.isLoading,
                isAdded: false,
                onClickAdd: { key, value in
                    viewModel.onEvent(event: OriginalRouteEvent.UpsertOriginalHeader(key: key, value: value))
                },
                onClickRemove: { _ in }
            )
            .padding(.horizontal)
        }
    }
    
    private var queriesSection: some View {
        Group {
            Text("Original Queries")
                .font(.headline)
                .foregroundColor(Color.primaryColor)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal)
            
            // Existing queries
            ForEach(Array(viewModel.state.originalQueries.keys.sorted()), id: \.self) { key in
                MapElement(
                    key: key,
                    value: viewModel.state.originalQueries[key] ?? "",
                    isLoading: viewModel.state.isLoading,
                    isAdded: true,
                    onClickAdd: { key, value in
                        viewModel.onEvent(event: OriginalRouteEvent.UpsertOriginalQuery(key: key, value: value))
                    },
                    onClickRemove: { key in
                        viewModel.onEvent(event: OriginalRouteEvent.RemoveOriginalQuery(key: key))
                    }
                )
                .padding(.horizontal)
            }
            
            // New query input
            MapElement(
                isLoading: viewModel.state.isLoading,
                isAdded: false,
                onClickAdd: { key, value in
                    viewModel.onEvent(event: OriginalRouteEvent.UpsertOriginalQuery(key: key, value: value))
                },
                onClickRemove: { _ in }
            )
            .padding(.horizontal)
        }
    }
    
    private var testSection: some View {
        TestColumn(
            isLoading: viewModel.state.isLoading,
            result: result.isEmpty ? "" : "Code: \(code)\nBody:\n\(result)",
            onClick: {
                result = ""
                viewModel.onEvent(event: OriginalRouteEvent.TestOriginalRoute(
                    originalBaseUrl: originalBaseUrl,
                    originalPath: originalPath,
                    originalMethod: originalMethod,
                    originalBody: originalBody
                ))
            }
        )
        .padding(.horizontal)
    }
    
    private func setupScreen() {
        viewModel.startObserving()
        subscribeToEvents()
    }
    
    private func cleanup() {
        viewModel.dispose()
        eventSubscription?.dispose()
    }
    
    private func subscribeToEvents() {
        eventSubscription = viewModel.eventFlow.subscribe(
            onCollect: { event in
                DispatchQueue.main.async {
                    guard let event = event else { return }
                    handleUiEvent(event)
                }
            }
        )
    }
    
    private func handleUiEvent(_ event: OriginalRouteViewModel.UiEvent) {
        switch event {
        case let showError as OriginalRouteViewModel.UiEventShowError:
            errorMessage = showError.message
            showErrorAlert = true
            
        case is OriginalRouteViewModel.UiEventShowOriginalRouteSuccess:
            break
            
        case let showResult as OriginalRouteViewModel.UiEventShowResult:
            code = Int(showResult.code)
            result = showResult.result
        case is OriginalRouteViewModel.UiEventNavigateToNextStep:
            let oldFields = generateOldBodyFieldsFromJson(result)
            let args = CreateRouteArgs(
                originalResponse: result,
                originalBaseUrl: originalBaseUrl,
                originalPath: originalPath,
                originalMethod: originalMethod,
                originalBody: originalBody,
                originalQueries: viewModel.state.originalQueries,
                originalHeaders: viewModel.state.originalHeaders,
                oldBodyFields: oldFields
            )
            navigate(.fillRouteFields(args))
        default:
            break
        }
    }
    
    // Add this JSON parsing function
    private func generateOldBodyFieldsFromJson(_ jsonString: String) -> [String: OldBodyField] {
        guard let data = jsonString.data(using: .utf8) else { return [:] }
        
        do {
            let json = try JSONSerialization.jsonObject(with: data)
            var fields: [String: OldBodyField] = [:]
            processJsonElement(json, fields: &fields, parents: [])
            return fields
        } catch {
            print("JSON parsing error: \(error)")
            return [:]
        }
    }
    
    private func processJsonElement(_ element: Any, fields: inout [String: OldBodyField], parents: [String]) {
        if let dict = element as? [String: Any] {
            for (key, value) in dict {
                if let nestedDict = value as? [String: Any] {
                    processJsonElement(nestedDict, fields: &fields, parents: parents + [key])
                } else if let array = value as? [Any], let firstObject = array.first as? [String: Any] {
                    processJsonElement(firstObject, fields: &fields, parents: parents + [key])
                } else {
                    let fieldName = parents.isEmpty ? key : (parents + [key]).joined(separator: "_")
                    let type: String
                    
                    switch value {
                    case is String: type = "String"
                    case is Int: type = "Int"
                    case is Double: type = "Double"
                    case is Bool: type = "Boolean"
                    default: type = "String" // fallback
                    }
                    
                    fields[fieldName] = OldBodyField(
                        keys: [key],
                        type: type,
                        parents: parents
                    )
                }
            }
        } else if let array = element as? [Any], let firstObject = array.first as? [String: Any] {
            processJsonElement(firstObject, fields: &fields, parents: parents)
        }
    }
}
